src/kinship/hierarchy-creator.ts

```diff
@@ -0,0 +1,303 @@
+    return new HierarchyCreator(data, startEntryId).createHierarchy();
+  }
+
+  readonly startEntryId: EntryId;  // Id of entry (indi or fam), which is root of the hierarchy
+  readonly startFamIndi: string;  // If startEntryId field is a fam id, then startFamIndi field can indicate which spouse in this family is the starting point of the hierarchy
+  readonly queuedNodesById = new Map<string, TreeNode>();
+  readonly idGenerator = new IdGenerator();
+
+  private constructor(readonly data: DataProvider<Indi, Fam>, startEntryId: EntryId) {
+    [this.startEntryId, this.startFamIndi] = this.expandStartId(startEntryId);
+  }
+
+  // Convert entry id to values of startEntryId and startFamIndi fields
+  private expandStartId(startEntryId: EntryId): [EntryId, string] {
+    if (!startEntryId) return [null, null];
+    if (startEntryId.isFam) return [startEntryId, null];
+    const indi = this.data.getIndi(startEntryId.id);
+    if (!indi) return [null, null];
+    const famsIds = indi.getFamiliesAsSpouse();
+    if (famsIds.length) return [EntryId.fam(famsIds[0]), startEntryId.id];
+    return [startEntryId, null];
+  }
+
+  createHierarchy(): Hierarchy {
+    const upRoot   = this.idToNode(this.startEntryId, null, null, false);
+    const downRoot = this.idToNode(this.startEntryId, null, null, false);
+    if (this.startFamIndi) {
+      upRoot.indi   = {id: this.startFamIndi};
+      downRoot.indi = {id: this.startFamIndi};
+    }
+    const layerEnd = {id: "LAYER END MARKER"} as TreeNode;
+    let [queue, otherQueue] = [[ upRoot, layerEnd ], [ downRoot ]];
+
+    while (queue.length) {
+      const node = queue.shift();
+      if (node !== layerEnd) {
+        const filter = node === upRoot || node === downRoot ?  //TODO: Filter only on root node?
+          node === upRoot ? HierarchyCreator.UP_FILTER : HierarchyCreator.DOWN_FILTER :
+          HierarchyCreator.ALL_ACCEPTING_FILTER;
+        this.fillNodeData(node, filter);
+        for (const childNode of node.childNodes.getAll()) {
+          queue.push(childNode);
+        }
+      } else {
+        [queue, otherQueue] = otherQueue.length ? [otherQueue, queue] : [queue, otherQueue];
+        if (queue.length) queue.push(layerEnd);
+      }
+    }
+
+    const getChildNodes = (node: TreeNode) => {
+      const childNodes = node.childNodes.getAll();
+      return childNodes.length ? childNodes : null;
+    };
+    return {
+      upRoot: d3.hierarchy(upRoot, getChildNodes),
+      downRoot: d3.hierarchy(downRoot, getChildNodes),
+    };
+  }
+
+  private fillNodeData(node: TreeNode, filter: HierarchyFilter) {
+    if (this.isFamNode(node)) {
+      const fam  = this.data.getFam(node.id);
+      const [indiId, spouseId] = (node.indi && node.indi.id === fam.getMother()) ?
+        [fam.getMother(), fam.getFather()] :
+        [fam.getFather(), fam.getMother()];
+      Object.assign(node, {
+        id: this.idGenerator.getId(node.id),
+        indi:   indiId   ? {id: indiId}   : null,
```

Comment:
You could write `indi: indiId && {id: indiId}` unless you want to make the `null` explicit.